JSON export and rehydration (issue 30)

qpwo opened the pull request:
Will add rehydration later

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt:
```diff
@@ -150,11 +151,12 @@ fun main(argv: Array<String>) {
+                        var jsonTree = ObjectMapper().readTree(jsonBytes) as ObjectNode
+                        jsonTree.replaceFieldEverywhere("sofaString", "__DOCUMENT_TEXT_REDACTED_FOR_IP_REASONS__")
+                        val outFileName = documentOutputDir.resolve("${document.name}-${annotationRecord.user}.json")
+                        Files.write(outFileName, jsonTree.toString().toByteArray())
```
In general, `toString` is to be avoided except for diagnostic purposes. Prefer to generate the JSON string using `ObjectMapper`.

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt:
```diff
@@ -150,11 +151,12 @@ fun main(argv: Array<String>) {
-                        val outFileName = documentOutputDir.resolve("${document.name}-${annotationRecord.user}.json");
-                        Files.write(outFileName, jsonBytes)
+                        var jsonTree = ObjectMapper().readTree(jsonBytes) as ObjectNode
+                        jsonTree.replaceFieldEverywhere("sofaString", "__DOCUMENT_TEXT_REDACTED_FOR_IP_REASONS__")
```
Add a comment about why we do this (that we cannot redistribute LDC documents and the user can always restore from those LDC documents)

rgabbard commented on pom.xml:
```diff
@@ -63,7 +63,7 @@
 				<groupId>edu.isi.vista</groupId>
 				<artifactId>common-core-open</artifactId>
-				<version>8.2.1-SNAPSHOT</version>
+				<version>8.2.2-SNAPSHOT</version>
```
What are we depending on which is not present in the release version?

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt:
```diff
@@ -100,38 +125,27 @@ fun main(argv: Array<String>) {
                     .resultObjectThrowingExceptionOnFailure<AeroResult<AnnotatorRecord>>(mapper)
                     .body
 
-            val documentOutputDir = projectOutputDir.resolve(document.name)
```
Why was this removed?

qpwo replied:
The output structure had one directory per file so we flattened it into just one directory per project. Should we revert that?

rgabbard replied:
How do we handle multiple annotators annotating the same file on the same project?

qpwo replied:
Oh is that possible? I thought each user has their own fresh copy of the project? (E.g. named `ArtifactExistence.DamageDestroy-bob_loblaw`)

rgabbard replied:
By policy, yes, but it is possible in the Inception data model for the someone else to come in an annotate on the project with someone else's name on it, so we should handle that case.

qpwo replied:
The containing dir names the project owner, and the file itself names the annotator. Does that eliminate possibility of collision? For example there is `output/ArtifactExistence.DamageDestroy-gabbard/AFP_ENG_19960918.0012-admin.json`

rgabbard replied:
Yep, that will do the trick. Maybe add a comment somewhere to clarify this?

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt:
```diff
@@ -172,21 +190,37 @@ private data class Document(val id: Long, val name: String, val state: String) {
+                "Request timed out. Waiting $timeoutInSeconds seconds and trying again."
+            }
+            TimeUnit.SECONDS.sleep(timeoutInSeconds)
+            this.retryOnFailure(maxTries - 1, timeoutInSeconds)
```
In this case, I think the simple for loop is clearer than recursion

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt:
```diff
@@ -172,21 +190,37 @@ private data class Document(val id: Long, val name: String, val state: String) {
+
+private fun Request.retryOnFailure(maxTries: Int = 3, timeoutInSeconds: Long = 30): ByteArray {
+    val (_, _, result) = this.response()
+    if (maxTries == 0)
```
Always use braces

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt:
```diff
@@ -0,0 +1,75 @@
+import java.io.File
+
+/**
+ * Puts the document text back into the 'sofaString' field inside json files created by `ExportAnnotations.kt`
```
Explain where and why they were removed

rgabbard replied:
Also mention this program in the repo's `README.md`

rgabbard replied:
Also, while we use the "dehydrate/rehydrate" language informally, it would be clearer to future readers if we called this something like `RestoreDocumentText`

rgabbard replied:
Still need to rename away from re-hydrate

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt:
```diff
@@ -0,0 +1,75 @@
+    val gigawordDataDirectory = params.getExistingDirectory("gigawordDataDirectory")
+    val jsonInputDirectory = params.getCreatableDirectory("dehydratedJsonDirectory")
+    val outputDirectory = params.getCreatableDirectory("rehydratedJsonDirectory")
+    outputDirectory.toString()
```
?

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt:
```diff
@@ -0,0 +1,75 @@
+ */
+
+fun main(argv: Array<String>) {
+    // Load parameters:
```
This comment doesn't add much :-)

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt:
```diff
@@ -0,0 +1,75 @@
+    outputDirectory.toString()
+
+    // Create an OriginalTextSource for getting original document text to put in json:
+    val textMap = DocIDToFileMappings.forFunction { symbol ->
```
You could consider pulling everything from here through lines 44 into its own method (possibly just taking the `Parameters` as an argument)

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt:
```diff
@@ -0,0 +1,75 @@
+            // First 21 characters of the filename are the document id
+            // Example filename is AFP_ENG_19960918.0012-admin.json
+            val docID = Symbol.from(jsonFile.name.substring(0, 21))
+            val text = textSource.getOriginalText(docID).get()
```
Check and raise an exception when `getOriginalText` returns `Optional.absent`. A possible idiom for this is:

```
val text = textSource.getOriginalText(docId).orNull() ?: 
       throw RuntimeException("Could not get original text for $docId")
```

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt:
```diff
@@ -0,0 +1,75 @@
+    }
+}
+
+fun docIDToFilename(symbol: Symbol?): String? {
```
Why do we accept `Symbol?` here rather than requiring `Symbol`?

qpwo replied:
`DocIDToFileMappings.forFunction` sends a `Symbol?` but I made `docIDToFilename` throw an error

qpwo commented on annotation-utils/pom.xml:
```diff
@@ -93,6 +93,11 @@
+		<dependency>
+			<groupId>edu.isi.vista</groupId>
+			<artifactId>nlp-core-open</artifactId>
+            <version>8.2.2-SNAPSHOT</version>
```
Deferring to @berquist here

berquist replied:
I think this was done in the struggle to get things compiling, where the crucial step was adding the dependency to the inner pom, not specifically making it be `-SNAPSHOT`. @qpwo Can you try making all of the `-SNAPSHOT`s into the previous stable version and see if it still works?

rgabbard replied:
8.2.2 has now been released, so you should be able to use 8.2.2 (without `-SNAPSHOT`) as the value now

berquist commented:
I think my commit https://github.com/isi-vista/curated-training-annotator/pull/33/commits/16bf9e11325879e5402f0552eaa31875cdca66b3 can be reverted (restore the deleted retry function and delete the recursive extension method) since we discovered the problem is with a particular annotation account on the server.

berquist commented on pom.xml:
```diff
@@ -63,12 +63,12 @@
 				<groupId>edu.isi.vista</groupId>
 				<artifactId>nlp-core-open</artifactId>
-				<version>8.1.0</version>
+				<version>8.2.2-SNAPSHOT</version>
```
Try depending on `8.2.1`.

berquist commented on pom.xml:
```diff
@@ -63,12 +63,12 @@
 				<groupId>edu.isi.vista</groupId>
 				<artifactId>common-core-open</artifactId>
-				<version>8.2.1-SNAPSHOT</version>
+				<version>8.2.2-SNAPSHOT</version>
```
Try depending on `8.2.1`.

qpwo replied:
Couldn't get it to work with 8.2.2. IntelliJ couldn't find it

rgabbard replied:
Does it work for 8.2.1?

rgabbard replied:
It looks like changes since 8.2.1 pertain only to `IndexFlatGigaword`: https://github.com/isi-vista/nlp-util/commits/master

qpwo replied:
Oops! I couldn't get it to work for 8.2.**1** but it works for 8.2.**2** so I think we should keep these changes.

rgabbard replied:
8.2.2 here as well, and below

berquist commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/Rehydrate.kt:
```diff
@@ -0,0 +1,81 @@
+fun makeTextSource(params: Parameters): OriginalTextSource {
+    // Create an OriginalTextSource for getting original document text to put in json:
+    val textMap = DocIDToFileMappings.forFunction { symbol ->
+        Optional.of(File(params.getExistingDirectory("gigawordDataDirectory"), docIDToFilename(symbol)))
```
Why is this `Optional`?

qpwo replied:
`DocIDToFileMappings.forFunction` wouldn't take it non-optionally and I couldn't get an `as`-cast to work either

rgabbard replied:
It expects a function which returns `Optional` because in general "I can't map this" is a valid response from a mapping strategy.

qpwo commented:
Did some more testing and indeed `8.2.2-SNAPSHOT` is required

rgabbard commented on annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt:
```diff
@@ -58,7 +58,9 @@ fun main(argv: Array<String>) {
     logger.info { "Connecting to Inception at $inceptionUrl" }
 
-    val mapper = ObjectMapper().registerKotlinModule()
+    val objectMapper = ObjectMapper()
```
Can you return this to its original form?  It is more idiomatic Jackson code; otherwise the reader is going to say "why did they do that?"

qpwo commented:
I get this error when I use 8.2.2 without the SNAPSHOT and run `ExportAnnotations.kt`:
```
10:32:32.808 [main] WARN  mu.KLogging - com.github.kittinunf.fuel.core.FuelError: Cannot construct instance of `edu.isi.vista.annotationutils.Document` (no Creators, like default construct, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: (BufferedInputStream); line: 1, column: 25] (through reference chain: edu.isi.vista.annotationutils.DocumentResult["body"]->java.util.ArrayList[0])
	com.github.kittinunf.fuel.core.FuelError$Companion.wrap(FuelError.kt:85)
	com.github.kittinunf.fuel.core.DeserializableKt.response(Deserializable.kt:176)
```

I think this statement is causing it:
```
1.        val documents = retryOnFuelError {
2.            getDocsUrl.httpGet(
3.                    parameters = listOf("projectId" to project.id))
4.                    .authenticateToInception()
5.                    .resultObjectThrowingExceptionOnFailure<DocumentResult>(mapper)
6.                    .body
7.        }
```

Where line 5 is the problem.

But why wouldn't this happen on the SNAPSHOT version? :thinking:

rgabbard commented:
@qpwo : Are you sure the version of the `nlp-utils` dependencies reliably causes this difference?  If you look at  https://github.com/isi-vista/nlp-util/commits/master, you'll see that (unless you have local modifications to `nlp-utils`):

* `8.2.2` and `8.2.2-SNAPSHOT` are identical
* `8.2.1` and `8.2.2` differ only in code internal to `IndexFlatGigaword`, which should be irrelevant here
* `8.2.1` differs from `8.2.0` only in code internal to `Parameters`, which should be irrelevant here.\\

qpwo commented:
You're right; I changed it back to `8.2.2-SNAPSHOT` and I got the same error as before. I changed it back to `8.2.2` and then did a `mvn clean install` and that killed the error somehow. Some sort of problem with changing dependencies in IntelliJ perhaps?

rgabbard commented:
@qpwo : Had you edited the dependencies in IntelliJ's interface directly or only via the `pom.xml`?

qpwo commented:
`pom.xml`

rgabbard commented:
@qpwo : If the problem is not present on a clean build, let's not worry about it for now.

qpwo commented:
Yes I agree I committed the changes earlier

rgabbard commented:
@qpwo : Is this PR ready for re-review?

qpwo commented:
Yes!

berquist commented:
I'm now getting lots of `FuelError` messages, and things basically worked before except for the one problem with https://github.com/isi-vista/curated-training-annotator/issues/34. Something happened between https://github.com/isi-vista/curated-training-annotator/pull/33/commits/5ea83801d0345d3e173cba5a067370e550bedd3b and the PR head.

rgabbard commented:
@qpwo : Poke me for re-review once you and @berquist have sorted out the problem above

qpwo commented:
@berquist Are you still getting the error after a `mvn clean install` and then `invalidate caches and restart` in IntelliJ?

qpwo commented:
Might be ready for a merge I'm still not getting errors
